### PR TITLE
Recsys novel

### PR DESCRIPTION
**Pull request by brahmaneya**
Add Recsys tutorial

The recsys test command is in the recsys tox env instead of the default, because it takes very long to run,  so we don't want to run it every time we run tox.

### CODE REVIEW

File: tox.ini [hunk 1]
```diff
@@ -17,6 +18,7 @@ deps =
     mtl: -rmtl/requirements.txt
     scene_graph: -rscene_graph/requirements.txt
     crowdsourcing: -rcrowdsourcing/requirements.txt
+    recsys: -rrecsys/requirements.txt
```
[the remark this answers is not in the document]
Comment (brahmaneya) [reply tox.ini]: I left it out on purpose, because this notebook takes a very long time to run (the dataset is huge).

File: recsys/recsys_tutorial.py [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+label_model.fit(L_train, n_epochs=5000, seed=123, log_freq=20, lr=0.01)
+
+from snorkel.analysis import metric_score
+from snorkel.utils import probs_to_preds
+
+Y_dev_prob = label_model.predict_proba(L_dev)
+Y_dev_pred = probs_to_preds(Y_dev_prob)
+
+acc = metric_score(data_dev.rating, Y_dev_pred, probs=None, metric="accuracy")
+print(f"LabelModel Accuracy: {acc:.3f}")
```
[the remark this answers is not in the document]
Comment (brahmaneya) [reply recsys/recsys_tutorial.py]: I actually got rid of the LabelModel accuracy altogether. It is misleading because our LFs have very low coverage (together they cover maybe 5% of the dataset). That is still enough training data for the model because the train set is so huge, but the LabelModel metrics are dominated by it's values on the uncovered examples.

File: recsys/recsys_tutorial.py [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+# %%
+import numpy as np
+import tensorflow as tf
+from tensorflow.keras import backend as K
+
+n_books = len(df_books)
+
+
+# Define recall, precision, and f1 score metrics to evaluate the model.
+def recall_m(y_true, y_pred):
```
[the remark this answers is not in the document]
Comment (brahmaneya) [reply recsys/recsys_tutorial.py]: They are operating on tf.Tensors, and y_pred is a probability instead of a boolean.
Moved all three to the utils.py file.

File: tox.ini (hunk 1 above)
[the remark this answers is not in the document]
Comment (brahmaneya) [reply tox.ini]: Need to figure out how to deal with this.

[the remark this answers is not in the document]
Comment (brahmaneya) [reply tox.ini]: I added the test command in a separate environment, so we can test recsys with tox -e recsys (and sync with tox -e recsys -- sync) but it won't run when you just run tox.

File: recsys/recsys_tutorial.py [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+combined_data_train = pd.concat([data_train_filtered, data_dev], axis=0)
+
+# %% [markdown]
+# ### Rating Prediction Model
+# We write a Keras model for predicting ratings given a user's book list and a book (which is being rated). The model represents the list of books the user interacted with, `books_idxs`, by learning an embedding for each idx, and averaging the embeddings in `book_idxs`. It learns another embedding for the `book_idx`, the book to be rated. Then it concatenates the two embeddings and uses an [MLP](https://en.wikipedia.org/wiki/Multilayer_perceptron) to compute the probability of the `rating` being 1.
+
+# %%
+import numpy as np
+import tensorflow as tf
+from utils import precision, recall, f1
```
[the remark this answers is not in the document]
Comment (brahmaneya) [reply recsys/recsys_tutorial.py]: They removed it, because the metrics are computed per batch, which is not really accurate.

File: recsys/recsys_tutorial.py [hunk 5]
```diff
@@ -300,3 +303,10 @@ def get_data_tensors(df):
 # %%
 test_data_tensors = get_data_tensors(data_test)
 model.evaluate(test_data_tensors[:-1], test_data_tensors[-1], steps=30)
+
+# %% [markdown]
+# ## Summary
+#
+# In this tutorial, we showed one way to use Snorkel for Recommendations. We used books metadata and review text to create `LF`s that estimate user ratings. We used a `LabelModel` to combine the outputs of those `LF`s. Finally, we trained a model to predict what books a user will read and like (and therefore what books should be recommended to the user) based only on what books the user has interacted with in the past.
```
Comment (bhancock8) [inline recsys/recsys_tutorial.py]: - no caps on Recommendations
- books'

File: recsys/recsys_tutorial.py [hunk 6]
```diff
@@ -300,3 +303,10 @@ def get_data_tensors(df):
...
+#
+# Note that we just focused on a specific setting in this tutorial; but Snorkel can also be used in different settings, where we have user profile data, where users have rated every book they read, and so on.
```
Comment (bhancock8) [inline recsys/recsys_tutorial.py]: Here we demonstrated one way to use Snorkel for training a recommender system. Note, however, that this approach could be easily adapted to take advantage of additional information as it is available (e.g., in the presence of user profile data, with denser user ratings, etc.).
